docs/extend/plugins_mountpoint.md

```diff
@@ -0,0 +1,214 @@
+the mount being requested. The plugin may allow or deny the mount,
+change the mount location, or perform arbitrary actions to setup the
+mount.
+
+<!-- TODO flow diagrams? -->
+
+## Docker client flows
+
+To enable and configure the mount point plugin, the plugin developer must
+support the Docker client interactions detailed in this section.
+
+### Setting up Docker daemon
+
+Enable the mount point plugin with a dedicated command line flag in the
+`--mount-point-plugin=PLUGIN_ID` format. The flag supplies a `PLUGIN_ID`
+value. This value can be the plugin’s socket or a path to a specification file.
+Mount point plugins can be loaded without restarting the daemon. Refer
+to the [`dockerd` documentation](../reference/commandline/dockerd.md#configuration-reloading) for more information.
+
+```bash
+$ dockerd --mount-point-plugin=plugin1 --mount-point-plugin=plugin2,...
+```
+
+Docker's mount point subsystem supports multiple `--mount-point-plugin` parameters.
+
+<!-- TODO request/response examples? -->
+
+## API schema and implementation
+
+In addition to Docker's standard plugin registration method, each plugin
+should implement the following three methods:
+
+* `/MountPointPlugin.MountPointProperties` The mount point plugin
+  properties method is called when mount point plugins are loaded at
+  daemon start- or reload-time. If containers are running which used a
+  now unloaded plugin during attachment, this method may also be called
+  before detachment if the plugin hasn't been initialized in the current
+  daemon instance (e.g. when using `--live-restore` and removing mount
+  point plugins while containers are running). The response to
+  `/MountPointPlugin.MountPointProperties` contains mount point filter
+  definitions which can be used to select which mount points (if any)
+  should be passed to this mount point plugin. This functionality
+  reduces container start-up latency even when many mount point plugins
+  may be in use.
+
+* `/MountPointPlugin.MountPointAttach` This mount point attachment method
+  is called before the Docker daemon starts a container. All applicable
+  mounts into the container are passed to the plugin simultaneously.
+
+* `/MountPointPlugin.MountPointDetach` This mount point detachment method
+  is called before the Docker daemon reports the shutdown status of a
+  container. Only the ID of the container causing the detachment is provided.
+
+#### /MountPointPlugin.MountPointProperties
+
+**Request**:
+
+```json
+{
+}
+```
+
+**Response**:
+
+```json
+{
+    "Success": "Indicates whether the properties query was successful (bool)",
+    "Types":   "Enumerates the types of mount points ('bind', 'volume') this mount point plugin interposes ({ string -> bool })"
+    "VolumePatterns": [
+        {
+            "VolumePlugin":  "A volume plugin name to require (string)",
+            "OptionPattern": "All of the keys of the OptionPattern must
```

Comment:
Yes, the filters need to be more expressive than they currently are and packing expressions into strings is not really appropriate (was it ever?). I'm working on this now.